sql: TestPrimaryKeyChangeWithOperations failed

sql.TestPrimaryKeyChangeWithOperations [failed](https://tanzanite.cluster.engflow.com/invocations/default/fcd1434c-2b66-4a39-ae86-da533a25f490?testReportRun=1&testReportShard=15&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [82b1a2b174dcedfd23b0a3a40c6069143f4b038f](https://github.com/cockroachdb/cockroach/commits/82b1a2b174dcedfd23b0a3a40c6069143f4b038f):


```
=== RUN   TestPrimaryKeyChangeWithOperations
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestPrimaryKeyChangeWithOperations54679927
    test_log_scope.go:76: use -show-logs to present logs inline
    schema_changer_test.go:344: schema change ALTER TABLE t.test ALTER PRIMARY KEY USING COLUMNS (k) took 578.138411ms
    schema_changer_test.go:2499: pq: duplicate key value violates unique constraint "test_x_key"
    schema_changer_test.go:2574: pq: index "test_x_key" not found
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestPrimaryKeyChangeWithOperations54679927
--- FAIL: TestPrimaryKeyChangeWithOperations (2.23s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=15</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPrimaryKeyChangeWithOperations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-47179

sql.TestPrimaryKeyChangeWithOperations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18761715?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18761715?buildTab=artifacts#/) on master @ [0563ba6941e3c5409aac567f9f806c2f91303d7f](https://github.com/cockroachdb/cockroach/commits/0563ba6941e3c5409aac567f9f806c2f91303d7f):


```
=== RUN   TestPrimaryKeyChangeWithOperations
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestPrimaryKeyChangeWithOperations3656270721
    test_log_scope.go:76: use -show-logs to present logs inline
    schema_changer_test.go:342: pq: duplicate key value violates unique constraint "crdb_internal_index_2_name_placeholder"
    schema_changer_test.go:344: schema change ALTER TABLE t.test ALTER PRIMARY KEY USING COLUMNS (k) took 416.73211ms
    schema_changer_test.go:2499: pq: duplicate key value violates unique constraint "test_pkey"
    schema_changer_test.go:2574: pq: index "test_x_key" not found
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestPrimaryKeyChangeWithOperations3656270721
--- FAIL: TestPrimaryKeyChangeWithOperations (4.87s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPrimaryKeyChangeWithOperations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql.TestPrimaryKeyChangeWithOperations [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18786240?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18786240?buildTab=artifacts#/) on master @ [85ecd9c5a3a91973185153fcff652cb7d84a1955](https://github.com/cockroachdb/cockroach/commits/85ecd9c5a3a91973185153fcff652cb7d84a1955):


```
=== RUN   TestPrimaryKeyChangeWithOperations
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestPrimaryKeyChangeWithOperations3328676538
    test_log_scope.go:76: use -show-logs to present logs inline
    schema_changer_test.go:342: pq: duplicate key value violates unique constraint "crdb_internal_index_2_name_placeholder"
    schema_changer_test.go:344: schema change ALTER TABLE t.test ALTER PRIMARY KEY USING COLUMNS (k) took 379.689567ms
    schema_changer_test.go:2626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestPrimaryKeyChangeWithOperations3328676538
--- FAIL: TestPrimaryKeyChangeWithOperations (2.12s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #140954 sql: TestPrimaryKeyChangeWithOperations failed [C-test-failure O-robot T-sql-foundations branch-release-25.1 release-blocker]
- #140709 sql: TestPrimaryKeyChangeWithOperations failed [C-test-failure O-robot T-sql-foundations branch-release-25.1.0-rc release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPrimaryKeyChangeWithOperations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -1730,9 +1730,15 @@ func (m *Manager) StartRefreshLeasesTask(ctx context.Context, s *stop.Stopper, d
 				// will be acquired if the schema is already leased to invalidate metadata
 				// caches (like optimizer memos).
 				purgeOldVersionsOrAcquireInitialVersion := func(ctx context.Context) {
+					if m.testingKnobs.TestingOnNewVersion != nil {
+						m.testingKnobs.TestingOnNewVersion(desc.GetID())
+					}
 					// Notify of any new / modified descriptors below once a new lease is
 					// acquired.
 					defer m.leaseGeneration.Add(1)
+					if m.testingKnobs.TestingOnLeaseGenerationBumpForNewVersion != nil {
+						defer m.testingKnobs.TestingOnLeaseGenerationBumpForNewVersion(desc.GetID())
+					}
 
 					// Whenever a new relation / type is created under an already leased
 					// schema we are going to lease the object out immediately. This allows

--- a/pkg/sql/catalog/lease/testutils.go
+++ b/pkg/sql/catalog/lease/testutils.go
@@ -50,6 +50,14 @@ type ManagerTestingKnobs struct {
 	// the lease manager attempts to acquire a lease for descriptor `id`.
 	TestingBeforeAcquireLeaseDuringRefresh func(id descpb.ID) error
 
+	// TestingOnNewVersion invoked when the range feed detects a new descriptor.
+	TestingOnNewVersion func(id descpb.ID)
+
+	// TestingOnLeaseGenerationBumpForNewVersion invoked when the lease generation,
+	// after a new descriptor or initial descriptor version are observed via
+	// the range feed.
+	TestingOnLeaseGenerationBumpForNewVersion func(id descpb.ID)
+
 	// To disable the deletion of orphaned leases at server startup.
 	DisableDeleteOrphanedLeases bool
 

--- a/pkg/sql/opt/cat/catalog.go
+++ b/pkg/sql/opt/cat/catalog.go
@@ -49,6 +49,11 @@ type StableID uint64
 // always return this value as their ID.
 const DefaultStableID = StableID(catid.InvalidDescID)
 
+// InvalidVersion is the version number used to represent objects that
+// are not versioned in any way. For example the GlobalPrivilege object,
+// which is not backed by a descriptor.
+const InvalidVersion = uint64(0)
+
 // SchemaName is an alias for tree.ObjectNamePrefix, since it consists of the
 // catalog + schema name.
 type SchemaName = tree.ObjectNamePrefix
@@ -260,8 +265,8 @@ type Catalog interface {
 	GetDependencyDigest() DependencyDigest
 
 	// LeaseByStableID leases out a descriptor by the stable ID, which will block
-	// schema changes. The underlying schema object is not returned.
-	LeaseByStableID(ctx context.Context, id StableID) error
+	// schema changes. The version of the underlying object is returned.
+	LeaseByStableID(ctx context.Context, id StableID) (uint64, error)
 
 	// GetRoutineOwner returns the username.SQLUsername of the routine's
 	// (specified by routineOid) owner.

--- a/pkg/sql/opt/cat/object.go
+++ b/pkg/sql/opt/cat/object.go
@@ -28,4 +28,7 @@ type Object interface {
 	//
 	// Used for invalidating cached plans.
 	Equals(other Object) bool
+
+	// Version returns the underlying version of the descriptor backing this object.
+	Version() uint64
 }

--- a/pkg/sql/opt/exec/explain/plan_gist_factory.go
+++ b/pkg/sql/opt/exec/explain/plan_gist_factory.go
@@ -498,6 +498,10 @@ func (u *unknownTable) ID() cat.StableID {
 	panic(errors.AssertionFailedf("not implemented"))
 }
 
+func (u *unknownTable) Version() uint64 {
+	panic(errors.AssertionFailedf("not implemented"))
+}
+
 func (u *unknownTable) PostgresDescriptorID() catid.DescID {
 	panic(errors.AssertionFailedf("not implemented"))
 }

--- a/pkg/sql/opt/indexrec/hypothetical_table.go
+++ b/pkg/sql/opt/indexrec/hypothetical_table.go
@@ -187,3 +187,8 @@ func (ht *HypotheticalTable) addInvertedCol(invertedSourceCol *cat.Column) *cat.
 	ht.invertedCols = append(ht.invertedCols, &invertedCol)
 	return &invertedCol
 }
+
+// Version is part of the cat.Object interface.
+func (ht *HypotheticalTable) Version() uint64 {
+	return 1
+}

--- a/pkg/sql/opt/metadata.go
+++ b/pkg/sql/opt/metadata.go
@@ -386,29 +386,44 @@ func (md *Metadata) dependencyDigestEquals(currentDigest *cat.DependencyDigest) 
 
 // leaseObjectsInMetaData ensures that all references within this metadata
 // are leased to prevent schema changes from modifying the underlying objects
-// excessively.
-func (md *Metadata) leaseObjectsInMetaData(ctx context.Context, optCatalog cat.Catalog) error {
-	for id := range md.dataSourceDeps {
-		if err := optCatalog.LeaseByStableID(ctx, id); err != nil {
-			return err
+// excessively. Additionally, the metadata version and leased descriptor versions
+// are compared.
+func (md *Metadata) leaseObjectsInMetaData(
+	ctx context.Context, optCatalog cat.Catalog,
+) (leasedVersionMatchesMetadata bool, err error) {
+	for id, ds := range md.dataSourceDeps {
+		ver, err := optCatalog.LeaseByStableID(ctx, id)
+		if err != nil {
+			return false, err
+		}
+		if ver != ds.Version() {
+			return false, nil
 		}
 	}
-	for id := range md.routineDeps {
-		if err := optCatalog.LeaseByStableID(ctx, id); err != nil {
-			return err
+	for id, rd := range md.routineDeps {
+		ver, err := optCatalog.LeaseByStableID(ctx, id)
+		if err != nil {
+			return false, err
+		}
+		if ver != rd.overload.Version {
+			return false, nil
 		}
 	}
-	for oid := range md.userDefinedTypes {
-		id := typedesc.UserDefinedTypeOIDToID(oid)
+	for _, typ := range md.userDefinedTypesSlice {
+		id := typedesc.UserDefinedTypeOIDToID(typ.Oid())
 		// Not a user defined type.
 		if id == catid.InvalidDescID {
 			continue
 		}
-		if err := optCatalog.LeaseByStableID(ctx, cat.StableID(id)); err != nil {
-			return err
+		ver, err := optCatalog.LeaseByStableID(ctx, cat.StableID(id))
+		if err != nil {
+			return false, err
+		}
+		if ver != uint64(typ.TypeMeta.Version) {
+			return false, nil
 		}
 	}
-	return nil
+	return true, nil
 }
 
 // CheckDependencies resolves (again) each database object on which this
@@ -440,8 +455,9 @@ func (md *Metadata) CheckDependencies(
 		// Lease the underlying descriptors for this metadata. If we fail to lease
 		// any descriptors attempt to resolve them by name through the more expensive
 		// code path below.
-		if err := md.leaseObjectsInMetaData(ctx, optCatalog); err == nil {
-			return true, nil
+		upToDate, err = md.leaseObjectsInMetaData(ctx, optCatalog)
+		if err == nil {
+			return upToDate, err
 		}
 	}
 

--- a/pkg/sql/opt/testutils/testcat/test_catalog.go
+++ b/pkg/sql/opt/testutils/testcat/test_catalog.go
@@ -487,8 +487,8 @@ func (tc *Catalog) GetDependencyDigest() cat.DependencyDigest {
 }
 
 // LeaseByStableID does not do anything since no leasing is used here.
-func (tc *Catalog) LeaseByStableID(_ context.Context, _ cat.StableID) error {
-	return nil
+func (tc *Catalog) LeaseByStableID(ctx context.Context, id cat.StableID) (uint64, error) {
+	return 1, nil
 }
 
 // ExecuteMultipleDDL parses the given semicolon-separated DDL SQL statements
@@ -690,6 +690,11 @@ func (s *Schema) ID() cat.StableID {
 	return s.SchemaID
 }
 
+// Version is a part of cat.Object
+func (s *Schema) Version() uint64 {
+	return 1
+}
+
 // PostgresDescriptorID is part of the cat.Object interface.
 func (s *Schema) PostgresDescriptorID() catid.DescID {
 	return catid.DescID(s.SchemaID)
@@ -746,6 +751,11 @@ func (tv *View) String() string {
 // ID is part of the cat.DataSource interface.
 func (tv *View) ID() cat.StableID {
 	return tv.ViewID
+}
+
+// Version is a part of cat.Object
+func (tv *View) Version() uint64 {
+	return 1
 }
 
 // PostgresDescriptorID is part of the cat.Object interface.
@@ -866,6 +876,11 @@ func (tt *Table) SetMultiRegion(val bool) {
 // ID is part of the cat.DataSource interface.
 func (tt *Table) ID() cat.StableID {
 	return tt.TabID
+}
+
+// Version is a part of cat.Object
+func (tt *Table) Version() uint64 {
+	return 1
 }
 
 // PostgresDescriptorID is part of the cat.Object interface.
@@ -1759,6 +1774,11 @@ var _ cat.Sequence = &Sequence{}
 // ID is part of the cat.DataSource interface.
 func (ts *Sequence) ID() cat.StableID {
 	return ts.SeqID
+}
+
+// Version is a part of cat.Object
+func (ts *Sequence) Version() uint64 {
+	return 1
 }
 
 // PostgresDescriptorID is part of the cat.Object interface.

--- a/pkg/sql/opt_catalog.go
+++ b/pkg/sql/opt_catalog.go
@@ -118,6 +118,11 @@ func (os *optSchema) ID() cat.StableID {
 	return cat.StableID(os.PostgresDescriptorID())
 }
 
+// Version is part of the cat.Object interface.
+func (os *optSchema) Version() uint64 {
+	return uint64(os.schema.GetVersion())
+}
+
 // PostgresDescriptorID is part of the cat.Object interface.
 func (os *optSchema) PostgresDescriptorID() catid.DescID {
 	switch os.schema.SchemaKind() {
@@ -518,11 +523,15 @@ func (oc *optCatalog) GetCurrentUser() username.SQLUsername {
 }
 
 // LeaseByStableID is part of the cat.Catalog interface.
-func (oc *optCatalog) LeaseByStableID(ctx context.Context, stableID cat.StableID) error {
+func (oc *optCatalog) LeaseByStableID(ctx context.Context, stableID cat.StableID) (uint64, error) {
 	// Lease the descriptor, so that schema changes cannot move forward
 	// after the current version.
-	_, err := oc.planner.byIDGetterBuilder().WithoutNonPublic().Get().Desc(ctx, descpb.ID(stableID))
-	return err
+	desc, err := oc.planner.byIDGetterBuilder().WithoutNonPublic().Get().Desc(ctx, descpb.ID(stableID))
+	if err != nil {
+		return 0, err
+	}
+	// Return the current version of the descriptor.
+	return uint64(desc.GetVersion()), nil
 }
 
 // GetDependencyDigest is part of the cat.Catalog interface.
@@ -687,6 +696,11 @@ func (ov *optView) ID() cat.StableID {
 	return cat.StableID(ov.desc.GetID())
 }
 
+// Version is part of the cat.Object interface.
+func (ov *optView) Version() uint64 {
+	return uint64(ov.desc.GetVersion())
+}
+
 // PostgresDescriptorID is part of the cat.Object interface.
 func (ov *optView) PostgresDescriptorID() catid.DescID {
 	return ov.desc.GetID()
@@ -758,6 +772,11 @@ func newOptSequence(desc catalog.TableDescriptor) *optSequence {
 // ID is part of the cat.Object interface.
 func (os *optSequence) ID() cat.StableID {
 	return cat.StableID(os.desc.GetID())
+}
+
+// Version is part of the cat.Object interface.
+func (os *optSequence) Version() uint64 {
+	return uint64(os.desc.GetVersion())
 }
 
 // PostgresDescriptorID is part of the cat.Object interface.
@@ -1191,6 +1210,11 @@ func newOptTable(
 // ID is part of the cat.Object interface.
 func (ot *optTable) ID() cat.StableID {
 	return cat.StableID(ot.desc.GetID())
+}
+
+// Version is part of the cat.Object interface.
+func (ot *optTable) Version() uint64 {
+	return uint64(ot.desc.GetVersion())
 }
 
 // PostgresDescriptorID is part of the cat.Object interface.
@@ -2399,6 +2423,11 @@ func newOptVirtualTable(
 // ID is part of the cat.Object interface.
 func (ot *optVirtualTable) ID() cat.StableID {
 	return ot.id
+}
+
+// Version is part of the cat.Object interface.
+func (ot *optVirtualTable) Version() uint64 {
+	return uint64(ot.desc.GetVersion())
 }
 
 // PostgresDescriptorID is part of the cat.Object interface.

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -7795,3 +7795,55 @@ func TestConcurrentDropAndCreateTable(t *testing.T) {
 	}
 	require.NoError(t, grp.Wait())
 }
+
+// TestLeaseGenerationBumpWithSchemaChange validates that DML queries properly
+// have plans invalidated even if the lease generation is bumped a bit later.
+// i.e. The lease generation is not a replacement for comparing versions, since
+// a new descriptor can be available slightly earlier then the generation bump.
+func TestLeaseGenerationBumpWithSchemaChange(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	ctx := context.Background()
+
+	descIDToDelay := descpb.InvalidID
+	grp := ctxgroup.WithContext(ctx)
+	var startDelayCallback func() chan struct{}
+	s, sqlDB, kvDB := serverutils.StartServer(t, base.TestServerArgs{
+		Knobs: base.TestingKnobs{
+			SQLLeaseManager: &lease.ManagerTestingKnobs{
+				TestingOnNewVersion: func(id descpb.ID) {
+					if id == descIDToDelay {
+						<-startDelayCallback()
+					}
+				},
+				TestingOnLeaseGenerationBumpForNewVersion: func(id descpb.ID) {
+					if id == descIDToDelay {
+						<-startDelayCallback()
+					}
+				},
+			},
+		},
+	})
+	defer s.Stopper().Stop(ctx)
+	runner := sqlutils.MakeSQLRunner(sqlDB)
+
+	var nextValue atomic.Int64
+	startDelayCallback = func() chan struct{} {
+		ch := make(chan struct{})
+
+		grp.GoCtx(func(ctx context.Context) error {
+			defer close(ch)
+			_, err := sqlDB.Exec("INSERT INTO t1 VALUES ($1, $2)", nextValue.Add(1), nextValue.Add(1))
+			return err
+		})
+
+		return ch
+	}
+
+	runner.Exec(t, "CREATE TABLE t1(n int not null, j int not null)")
+	runner.Exec(t, "INSERT INTO t1 VALUES ($1, $2)", nextValue.Add(1), nextValue.Add(1))
+	tableDesc := desctestutils.TestingGetPublicTableDescriptor(kvDB, keys.SystemSQLCodec, "defaultdb", "t1")
+	descIDToDelay = tableDesc.GetID()
+	runner.Exec(t, "ALTER TABLE t1 ALTER PRIMARY KEY USING COLUMNS(n, j)")
+	require.NoError(t, grp.Wait())
+}

--- a/pkg/sql/syntheticprivilege/global_privilege.go
+++ b/pkg/sql/syntheticprivilege/global_privilege.go
@@ -64,6 +64,11 @@ func (p *GlobalPrivilege) ID() cat.StableID {
 	return cat.DefaultStableID
 }
 
+// Version implements the cat.Object interface.
+func (p *GlobalPrivilege) Version() uint64 {
+	return cat.InvalidVersion
+}
+
 // PostgresDescriptorID implements the cat.Object interface.
 func (p *GlobalPrivilege) PostgresDescriptorID() catid.DescID {
 	return descpb.InvalidID

